### ISSUE TITLE
Configuration of DB connection encoding

### ISSUE THREAD

**Issue by pes-soft** [opencats/OpenCATS#44]
By default, new database connections via `mysql_connect` use *latin1* character encoding ( based on MySQL configuration ). This configuration works for standard functionality of OpenCATS, but it can degrade user experience in certain circumstances related to database ( i.e. search functionality, direct extraction of data from database by third party applications ).

For example this influences search of contacts, companies and other subjects which include special characters ( like á ) in the names. When DB connection is identical to database ( UTF-8 ), users can search for names with special characters without need to exactly specify the name ( for example *Libéreau* can be searched by *libereau* ).

Configuration of DB connection encoding could be provided in *config.php* and applied by `mysql_set_charset`, when connection is being established.

**Comment by RussH** [comment]
Agreed. Some time ago I changed the db encoding of the created databases in my instance. Suspect I'll change default encoding to utf8 for the standard install.

**Comment by pes-soft** [comment]
Great :thumbsup:

One more comment: `mysql_set_charset` should be called only in the case, when it has been enabled during new installation. Otherwise it should be avoided - so upgraded OpenCATS setup database connection the same way as before.

**Comment by RussH** [comment]
Hi, do you have some code to correct this, or I'll leave it unassigned for now?

**Comment by mlespiau** [comment]
I have done some changes to fix this problem, but have not thought of a good way for supporting old installations and new code. 

In my opinion we should force everyone to move to utf8, maybe we could include a great migration guide from previous version to the new version that can update the db from latin1 to utf8, but it will involve too much extra work.

Do we have any idea of how many opencats installations are out there running in production?

**Comment by amaisonneuve** [comment]
What the impact of this change? I haven't looked at this yet.

**Comment by mlespiau** [comment]
I have just created a PR: https://github.com/opencats/OpenCATS/pull/81

There are more utf8 issues reported but this is a good milestone towards it's resolution.

**Comment by RussH** [comment]
Opencats installations - I can't honestly say.

Maybe we can change the 'new version check' in opencats to hit the
opencats.org site going forward.

Looking at the code, it tries to hit www.catsone.com/catsnewversion.php:80
everytime a user logs on (unless the user has disabled the check - we've
set this as the default in 0.9.2)

However a quick google search shows a few hundred have the default 'career
portal' enabled.
--> Google search for (+"Powered by cats" +"Cognizo" -opencats.org -
github.com -catsone.com)

... however I suspect there's allot more that don't use it as an
internet-facing tool - for example (my wife's company) doesn't have the
internet-facing jobs-portal and the login screen is protected by htaccess -
so there's no way to detect it.







----
Russ

**Comment by skrchnavy** [comment]
Before making migration, it shall be chosen what are proper (or supported) collations.
`utf8` is transfered usualy to `utf8_general_ci`.
see https://mediatemple.net/community/products/dv/204403914/default-mysql-character-set-and-collation
One commoent to PR #81 - is it necessary to create tables with CHARSET? Isn;t it OK to create DB with proper collation and reuse it in all tables?

Regarding migration, in the page I linked above there is also a description how to migrate tables using phpmyadmin and also command line to new collation.
latin1 shall be 'forward' compatible with utf8 so it shall be withot loosing of data.

**Comment by mlespiau** [comment]
Great points @skrchnavy.

After some further research I think the right move is to go to [utf8mb4 for full unicode support](https://mathiasbynens.be/notes/mysql-utf8mb4).

Regarding the collation, I would rather use `_unicode_` than `_general_`. The main difference is in sorting:

- unicode sorts accurately in a very wide range of languages
- general fails to implement all of the Unicode sorting rules, which will result in undesirable sorting in some situations, such as when using particular languages or characters

It seems that general performs slightly better than unicode, but on modern servers, this performance boost will be all but negligible.

> One commoent to PR #81 - is it necessary to create tables with CHARSET? Isn;t it OK to create DB with proper collation and reuse it in all tables?
No it's not, doing it at DB level should be enough as tables should get the database charset by default.

> latin1 shall be 'forward' compatible with utf8 so it shall be withot loosing of data.
True, but beware, if the DB is not migrated and the config.php is left as utf8, PHP will talk to MySQL in utf8, not latin1. 
MySql does not validates the input and stores the data bytes as they come, leading to have UTF8 data in Latin1 DBs. There are tricks to restore those pieces of data but it's very manual and painful.

**Comment by pes-soft** [comment]
I think that using utf8mb4 with unicode collation for database is a good idea too.

How do you suggest to configure encoding for database connection? Should it be also forced to utf8mb4 ( at least for new installations ) or left for PHP to decide?

**Comment by mlespiau** [comment]
> How do you suggest to configure encoding for database connection? Should it be also forced to utf8mb4 ( at least for new installations ) or left for PHP to decide?

I would force it and indeed in master now it's being forced to 'SQL_CHARACTER_SET' constant in `config.php`. So we should change:
`define('SQL_CHARACTER_SET', 'utf8');`
with:
`define('SQL_CHARACTER_SET', 'utf8mb4');`

**Comment by RussH** [comment]
do you want to commit a change?



----
Russ

**Comment by mlespiau** [comment]
We're now using 'utf8' in master, we definitely should move to utf8mb4. It would be great if someone else could work on this and write a small migration guide. If that's not possible, I can do it but would like to finish the filtering functionality I'm working on first.
